# Parcial 1 · Machine Learning (Primer Corte)
**Formato:** Notebook de Python (Colab/Jupyter)

**Instrucciones**
- Responde directamente en las celdas indicadas con *Tu respuesta aquí*.
- No borres el enunciado de las preguntas.
- No uses gráficos; solo filtros, agrupaciones y cálculos básicos.
- Mantén el código y comentarios en **español**.
- **Entrega**: sube este `.ipynb` resuelto a la plataforma indicada.

**Dataset esperado (`students.csv`)**
Columnas: `gender, race/ethnicity, parental level of education, lunch, test preparation course, math score, reading score, writing score`.

**Estructura del parcial**
- **Sección A — Fundamentos (6 preguntas)**
- **Sección B — EDA con filtros y agrupaciones (6 preguntas)**

**Puntaje**: 100 puntos (≈ 8.3 por pregunta).

---

In [1]:
!wget https://raw.githubusercontent.com/javierherrera1996/lecture_analytics/main/StudentsPerformance.csv

--2025-09-17 20:48:33--  https://raw.githubusercontent.com/javierherrera1996/lecture_analytics/main/StudentsPerformance.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 72036 (70K) [text/plain]
Saving to: ‘StudentsPerformance.csv’

StudentsPerformance 100%[===================>]  70.35K  --.-KB/s    in 0.02s   

2025-09-17 20:48:33 (3.09 MB/s) - ‘StudentsPerformance.csv’ saved [72036/72036]



### Dependencias (ejecuta esta celda primero)

In [2]:
import pandas as pd
import numpy as np

## Carga del dataset
Sube un archivo **`students.csv`** (mismo directorio del notebook) o ajusta la ruta en la celda. Si no tienes el archivo, puedes **generar una muestra mínima** con la celda opcional.


In [3]:
df = pd.read_csv("StudentsPerformance.csv")

#### (Opcional) Generar una muestra mínima si no tienes el CSV

## Sección A — Fundamentos de Machine Learning (Conceptual, sin métricas ni gráficos)

### A1. Supervisado vs No Supervisado
**Describe** con tus palabras la diferencia entre **aprendizaje supervisado** y **no supervisado**. Da **1 ejemplo** de cada uno aplicado a educación o negocios.

_Tu respuesta aquí (Markdown):_

El aprendizaje supervisado es aquel aprendizaje que se basa en prueba y error por medio de etiquetas o respuestas que ya son conocidas por la computadora, ya que la computadora recibe una pregunta y respuesta que son correctas y con base a este puede predecir en  sobre nuevos casos. En la vida real esto se puede ver en los estudiantes, es decir tratar de predecir si un estudiante va a pasar el año o va a tener cierta nota especifica por el historia académico que el estudiante tenga, si el estudiante ha tenido buenas calificaciones siempre, es muy probable que gane el año o tenga una nota muy alta, en los negocios es similar, puede ser las compras de un proveedor a una empresa en específico o las ventas de cierto producto, ya que el nivel de preferencia es mayor.

El aprendizaje no supervisado es al contrario que el supervisado, ya que aquí no hay etiquetas o modelos de preguntas y respuestas corectas si no por patrones, donde se van explorando datos para encontrar patrones que están ocultos, o agrupaciones que para la máquina es compleja de encontrar en un primer momento, de esta manera los computadores organizan la información y buscan varias fuentes para encontrar una respuesta óptima, un ejemplo de esto sería la segmentación de los clientes en una empresa para un producto de lanzamiento, la empresa hace un gasto promedio, la frecuencia de compra que tendría, el tipo de cliente o cliente ideal y con base a esto se hace una estrategia de marketing personalizada para el éxito del producto.

### A2. Variable dependiente e independientes
Si queremos predecir `math score`, ¿cuál es la **variable dependiente**? Menciona **dos** posibles **variables independientes** justificando por qué.

_Tu respuesta aquí (Markdown):_


mis variable serían:  test preparation course y parental level of education, ya que al tener padres con un nivel de educación superior, tiene la gran pocibilidad de que su hijo tenga una preparación previa a un exámen para que su hijo este más preparado y seguro de la prueba y así poder tener una nota superiro, a diferencia de tener unos padres sin la base o los conocimientos de estas pruebas o pensar de que no son necesario por el hecho de que no se han encontrado en una situación similar antes, hace que el niño probablemente tenga una preparación menor o no le de mucha importancia a la nota que adquiera o se conforme con poco.

### A3. Clasificación binaria
Si clasificamos estudiantes en **Aprobado** (≥ 60) vs **No aprobado**, ¿qué tipo de aprendizaje es? Justifica brevemente.

_Tu respuesta aquí (Markdown):_

Si queremos clafificar a los estudiantes en aprobado mayor a 60 es un lenguaje supervisado ya que hay una variable ya definida que es estudiantes aprobados mayor o igual a 60, teniendo esto, el sistema o el computador tiene una base y puede aprender a clasificar a los estudiantes con esta etiqueta definida claramente.


### A4. Casos de negocio (supervisado)
Da **dos** ejemplos de problemas de negocio donde usarías aprendizaje **supervisado** y explica por qué.

_Tu respuesta aquí (Markdown):_

La predicción de las ventas de una empresa, ya que se puede entrenar al sistema con una base de datos histórica que tenga la empresa, diferentes modelos de ventas anteriores, y que con esto genere estrategias de marketing y haga una predicción de las ventas que posiblemente puede llegar a  tener la empresa.

la otra sería riesgos de créditos, ya que aquí se pueden clasificar a los clientes como aprobado, rechazado junto con su historial crediticio y de esta manera ponerle condicionales y así poder hacer que el sistema haga un análisis de a que personas se puede prestar dinero y a quienes no y su tipo de riesgo.

### A5. Casos de negocio (no supervisado)
Da **dos** ejemplos de problemas de negocio donde usarías aprendizaje **no supervisado** y explica por qué.

_Tu respuesta aquí (Markdown):_

El primer problema de negocio que usaria para el aprendizaje no supervisado sería la recomendación de productos a los diferentes tipos de clientes que puede tener una tienda, dependiendo de la frecuencia de compra, la relación de los productos que adquiere la persona, se puede lograr que el sistema recomiende productos con base a etiquetas explicitas o específicas.

La otra opción es la optimización de rutas de logística, ya que si tenemos la información de donde son los puntos con mayores entregas o la frecuencia de los pedidos en ciertas áreas, se puede lograr descubrir agrupaciones para reducir los tiempos de entrega y reparto de los productos.

### A6. Segmentación por similitud
Si quieres segmentar estudiantes por similitud de `reading score` y `writing score`, ¿qué tipo de aprendizaje usarías y por qué?

_Tu respuesta aquí (Markdown):_

usaría el aprendizaje no supervisado porque a pesar de que el sistema tiene la restricción de reading score y writing score, tendría que hacer un análisis el solo, y detectar las similitudes o los patrones y hacer una agrupación entre los estudiates con semejanzas en el puntaje.

## Sección B — EDA (solo filtros y agrupaciones, foco de negocio)

### B7. Promedio de matemáticas por género
Calcula el **promedio de `math score` por `gender`**.
- ¿Qué género tiene mejor promedio?
- **Conclusión de negocio**: en 1–2 líneas, ¿qué harías con esta información?

_Escribe tu código y comentarios:_

In [4]:
df.groupby('gender')['math score'].mean()

gender
female    63.633205
male      68.728216
Name: math score, dtype: float64

El género que tiene mejor promedio son los Hombres, con esta información trataría de mirar si hay algún fallo con la comprensión de matemáticas en las chicas para así mejorar la metodología de clase para los dos géneros.

### B8. Lectura por grupo étnico
Calcula el **promedio de `reading score` por `race/ethnicity`**.
- ¿Qué grupo tiene mejor desempeño?
- **Aplicación**: ¿cómo usarías esto en un programa de apoyo escolar?

_Escribe tu código y comentarios:_

In [8]:
df.groupby('race/ethnicity')['reading score'].mean().sort_values(ascending=False)

race/ethnicity
group E    73.028571
group D    70.030534
group C    69.103448
group B    67.352632
group A    64.674157
Name: reading score, dtype: float64

El grupo que tiene mejor desempeño es el grupo E, con un desempeño promedio de 73%.
Con estos resultados, implementaría planes de acción para que los grpos con menor promedio puedan reforzar sus conocimientos, aprender tips o cosas nuevas que los impulse e iguale a los del grupo E, también una retroalimentación de los fallos que tuvieron y el por que lo tuvieron, para que así tengan una visión y comprensión mayor y puedan mejorar continuamente.

### B9. Impacto del tipo de almuerzo
Filtra estudiantes con `lunch = free/reduced` y calcula el **promedio de `math score` y `reading score`**. Compáralo con `lunch = standard`.
- **Conclusión de negocio**: ¿qué acción propones?

_Escribe tu código y comentarios:_

In [9]:
df.groupby('lunch')[['math score', 'reading score']].mean()

math score  reading score
lunch                                  
free/reduced   58.921127      64.653521
standard       70.034109      71.654264

Los gobiernos deben ofrecer comida de calidad a los estudiantes y más cuando se encuentran en su edad de crecimiento, ya que los niños necesitan más energía y más nutrientes para estar sanos y fuertes y no tengan ningún problema en su desarrollo.

### B10. Educación parental y rendimiento
Agrupa por `parental level of education` y calcula el **promedio de `math score`**.
- ¿Qué nivel se asocia con mejores resultados?
- **Acción**: ¿qué intervención o programa harías basado en esto?

_Escribe tu código y comentarios:_

In [10]:
df.groupby('parental level of education')['math score'].mean().sort_values(ascending=False)

parental level of education
master's degree       69.745763
bachelor's degree     69.389831
associate's degree    67.882883
some college          67.128319
some high school      63.497207
high school           62.137755
Name: math score, dtype: float64

El nivel que se asocia con los mejores resultados es el master's degree con un porcentaje del 69.7%

los programas que haría seria una escuela para padres, intentar capacitarlos para que puedan darle una mejor educación a sus hijos o mantenerlos actualizados para que puedan ayudar a abrirle las puertas a sus hijos y más con  los diferentes apoyos o recursos que tienen los gobiernos, a su vez, implementaría una formación gratuita y de calidad a las personas que no cuentan con padres presentes o que sus padres no tienen un nivel de educación superior, para que el niño no pierda las oportunidades que pueden tener los demás.

### B11. Curso de preparación y puntaje
Filtra estudiantes que **completaron** `test preparation course` y calcula el **promedio de `math score`**. Compáralo con quienes **no** lo completaron.
- **Decisión**: ¿financiarías este curso? Explica.

_Escribe tu código y comentarios:_

In [11]:
df.groupby('test preparation course')['math score'].mean()

test preparation course
completed    69.695531
none         64.077882
Name: math score, dtype: float64

Sí, ya que las personas que tomaron el curso tuvieron mejores resultados que las personas que no lo hiciern, esto ayudaría mucho a los estudiantes y beneficiaría en gran manera a aquellos que no tienen la posibilidad económica de completarlo o de adquirirlo por sus propios medios.

### B12. Doble agrupación (negocio)
Crea una **tabla** agrupada por `gender` y `test preparation course` con el **promedio de `writing score`**.
- ¿Qué combinación muestra mejores resultados?
- **Recomendación**: ¿a quién priorizas para una beca del curso?

_Escribe tu código y comentarios:_

In [13]:
df.groupby(['gender', 'test preparation course'])['writing score'].mean()


gender  test preparation course
female  completed                  78.793478
        none                       68.982036
male    completed                  69.793103
        none                       59.649351
Name: writing score, dtype: float64

La combinación que muestra mejores resultados son el grupo de las mujeres, pero priorizaría una beca para los hombres que no relizaron el preparatorio, ya que ellos son los que más necesitan refuerzo en este tema y les ayudaría mucho en el refuerzo y también merecen una oportunidad de también tener altos resultados, es decir, no siempre premiar a los mismos por sus resultados, si no darles la oportunidad a otras personas que no puedan o no tengan la posibilidad de hacerlo.